lib/auth/winwebauthn/webauthn_windows.go

```diff
@@ -0,0 +1,945 @@
+		cCredentials:  uint32(len(exCredList)),
+		ppCredentials: &exCredList[0],
+	}, nil
+}
+
+func transportsToCType(in []protocol.AuthenticatorTransport) uint32 {
+	if len(in) == 0 {
+		return 0
+	}
+	var out uint32
+	for _, at := range in {
+		// WEBAUTHN_CTAP_TRANSPORT_USB         0x00000001
+		// WEBAUTHN_CTAP_TRANSPORT_NFC         0x00000002
+		// WEBAUTHN_CTAP_TRANSPORT_BLE         0x00000004
+		// WEBAUTHN_CTAP_TRANSPORT_TEST        0x00000008
+		// WEBAUTHN_CTAP_TRANSPORT_INTERNAL    0x00000010
+		switch at {
+		case protocol.USB:
+			out += 0x1
+		case protocol.NFC:
+			out += 0x2
+		case protocol.BLE:
+			out += 0x4
+		case protocol.Internal:
+			out += 0x10
+		}
+	}
+	return out
+}
+
+func attachmentToCType(in protocol.AuthenticatorAttachment) uint32 {
+	// WEBAUTHN_AUTHENTICATOR_ATTACHMENT_ANY                               0
+	// WEBAUTHN_AUTHENTICATOR_ATTACHMENT_PLATFORM                          1
+	// WEBAUTHN_AUTHENTICATOR_ATTACHMENT_CROSS_PLATFORM                    2
+	switch in {
+	case protocol.Platform:
+		return 1
+	case protocol.CrossPlatform:
+		return 2
+	default:
+		return 0
+	}
+}
+
+func conveyancePreferenceToCType(in protocol.ConveyancePreference) uint32 {
+	// WEBAUTHN_ATTESTATION_CONVEYANCE_PREFERENCE_ANY                      0
+	// WEBAUTHN_ATTESTATION_CONVEYANCE_PREFERENCE_NONE                     1
+	// WEBAUTHN_ATTESTATION_CONVEYANCE_PREFERENCE_INDIRECT                 2
+	// WEBAUTHN_ATTESTATION_CONVEYANCE_PREFERENCE_DIRECT                   3
+	switch in {
+	case protocol.PreferNoAttestation:
+		return 1
+	case protocol.PreferIndirectAttestation:
+		return 2
+	case protocol.PreferDirectAttestation:
+		return 3
+	default:
+		return 0
+	}
+}
+
+func userVerificationToCType(in protocol.UserVerificationRequirement) uint32 {
+	// WEBAUTHN_USER_VERIFICATION_REQUIREMENT_ANY                          0
+	// WEBAUTHN_USER_VERIFICATION_REQUIREMENT_REQUIRED                     1
+	// WEBAUTHN_USER_VERIFICATION_REQUIREMENT_PREFERRED                    2
+	// WEBAUTHN_USER_VERIFICATION_REQUIREMENT_DISCOURAGED                  3
+	switch in {
+	case protocol.VerificationRequired:
+		return 1
+	case protocol.VerificationPreferred:
+		return 2
+	case protocol.VerificationDiscouraged:
+		return 3
+	default:
+		return 0
+	}
+}
+
+func requireResidentKeyToCType(in *bool) uint32 {
+	if in == nil {
+		return 0
+	}
+	return boolToUint32(*in)
+}
+
+func (c client) makeCredOptionsToCType(in protocol.PublicKeyCredentialCreationOptions) (*_WEBAUTHN_AUTHENTICATOR_MAKE_CREDENTIAL_OPTIONS, error) {
+	exCredList, err := credentialsExToCType(in.CredentialExcludeList)
+	if err != nil {
+		return nil, err
+	}
+
+	var dwVersion uint32
+	switch c.version {
+	case apiVersion1, apiVersion2:
+		dwVersion = 3
+	case apiVersion3:
+		dwVersion = 4
+	case apiVersion4:
+		dwVersion = 5
+	}
+
+	// TODO (tobiaszheller): version of duo-labs/webatuhn used in teleport does
+	// not support PreferResidentKey field. We cannot easily update version due
```

Comment:
Yes, for hello they are always resident, similary to TouchID implementation. I assume that this field is used for FIDO flow but this is shot because we are missing correct docs.

Reply:
Let's mention somewhere in docs and/or comments that Hello keys are always resident, that's good to know.

I have an ongoing update for duo-labs/webauthn, but we never set "preferred" anyway, so feel free to tweak the TODO/code accordingly.

Reply:
added comments and update MakeCredential func docs with info about resident keys.